### ISSUE TITLE
250827 Computer Vision Basic 장윤정

### ISSUE THREAD

**Issue by yjjang06** [AIBootcamp14/wiki-rider#38]
## 📅 학습 날짜  
2025-08-27

## 📌 키워드 및 요약 정리
**키워드 :** CNN, Convolution Layer, Batch Normalization, Activation Layer, Pooling Layer

### 🔶 1. Convolution Layer

Convolution Layer에서는 컴퓨터 비전 태스크를 수행하는데 유용한 피쳐들을 학습합니다.
- 하위 레이어에서는 Edge와 같은 low-level feature를 주로 학습
- 상위 레이어에서는 Shape과 같은 high-level feature를 주로 학습

#### 🔹 주요 개념

**필터(Filter)**
- 입력 이미지를 특정 크기의 필터를 이용하여 합성곱(Convolution)연산하여 필터 영역에 대한 특징을 추출합니다.
- 합성곱 커널(Kernel)이라고도 합니다.

**Stride**
- 필터가 입력 이미지 위에서 한 번에 이동하는 간격을 의미합니다.

**Padding**
- 특징 맵의 크기를 일정 수준으로 유지하기 위해 입력 이미지 가장자리에 0 또는 특정 값을 추가하는 것 입니다.
- 합성곱 연산 시 출력 크기 축소를 조절하거나 방지하기 위해 사용합니다.

**특징 맵(Feature Map)**
- 필터를 사용하여 이미지 내 특징들을 추출한 결과물 입니다.
- 각 필터는 이미지에서 특정 패턴을 감지하여 학습할 수 있게 하며, 필터의 개수는 모델이 학습한 패턴의 개수를 의미합니다.
- 각 필터에 대해 학습한 결과물인 특징 맵은 특정 위치에 해당 패턴이 있는지 없는지, 얼마나 강하게 존재하는지를 나타냅니다.

---

### 🔶 2. Batch Normalization

모델의 학습을 더 빠르고 안정적으로 만들기 위해 사용하며, Convolution Layer 다음, Activation Layer 전에 사용합니다.

신경망의 각 레이어는 학습 과정에서 계속해서 입력 데이터의 분포가 달라지는데,
이 경우, 학습이 진행될 수록 각 레이어의 가중치를 불안정하게 만들고, 최적화 과정을 방해하여 학습 속도를 늦춥니다.

이를 방지하기 위해 배치 정규화(Batch Normalization)를 사용하여 각 레이어의 입력 분포를 일정하게 유지합니다.

- 학습 분포가 항상 일정하게 유지되어 학습이 **안정적**으로 진행되고, 최적의 가중치를 **더 빠르게** 찾을 수 있습니다.
- 학습이 안정되기 때문에, **높은 학습률(Learning Rate)을 사용**할 수 있어 학습 속도를 더 가속화 합니다.
- 배치 정규화는 일종의 **정규화(Regularization)** 효과를 제공하여 **과적합(Overfitting)을 방지**할 수 있습니다.

---

### 🔶 3. Activation Layer

모델에 비선형성을 부여해주기 위해 사용합니다.
- 복잡한 문제를 해결하기 위해서는 비선형성을 지닌 모델이 유리합니다.

#### 🔹 주요 활성화 함수

**1. Sigmoid**
  - (0, 1) 사이의 값이 출력됩니다. (출력이 항상 양수)
  - 연속적이고, 모든 값에 대해 미분이 가능합니다.
  - 출력 값이 0 또는 1에 가까운 경우, 경사가 0에 가까우며 기울기 소실(Vanishing Gradient)이 발생할 수 있습니다.

**2. tanh** 
  - (-1, 1) 사이의 값이 출력됩니다.
  - 0을 중심으로 대칭되어 있습니다. (zero centroid)
  - Sigmoid와 마찬가지로 입력 값이 매우 크거나 작으면 기울기 소실이 발생합니다.

**3. ReLU**
  - 입력 값이 양수일 때, 기울기 소실이 발생하지 않는 활성화 함수입니다.
  - Sigmoid나 tanh보다 학습 속도가 빠릅니다.
  - 입력 값이 음수일 경우, gradient가 0이기 때문에 해당 뉴런은 학습되지 않는 문제가 발생할 수 있습니다.

**4. Leak ReLU**
  - 입력 값이 음수일 경우, ReLU 사용 시 gradient가 0이 되는 문제를 보완한 활성화 함수입니다.
  - 입력 값이 음수일 경우 0.1x, 양수일 경우 x값이 출력됩니다. 

---

### 🔶 4. Pooling Layer

특징 맵(Feature Map)의 공간 정보를 축소하면서 중요한 특징을 요약합니다. 이를 공간적 통합(Spatial Aggregation)이라고 합니다.
- 풀링 윈도우(Pooling window) 사이즈 기준으로 해당 영역 내 픽셀 정보를 통합하여 하나의 픽셀값으로 만들어줍니다.
- 출력 크기를 줄여 연산량을 감소시키고, 더 넓은 Receptive Field를 확보할 수 있습니다.

#### 🔹 대표방식

|   **Max Pooling** | **Average Pooling** |
| --- | --- | 
| - 풀링 윈도우 내에서 최대값을 취한다 | - 풀링 윈도우 내에서 평균값을 취한다 |
| - 강한 특징을 유지(강조)한다 | - 특정 영역의 전반적인 특성을 유지한다 |
| - 사소한 노이즈 제거 효과 | - 특징 맵의 전반적인 정보를 부드럽게 요약 |
| - 정보의 손실이 발생할 수 있다 | - 중요한 정보가 희석될 수 있다 |

## ❓ 어려웠던 부분  
- 상위 레이어로 갈수록 Receptive Field를 넓게 보고, 더 많은 특징들을 파악할 수 있다는 부분이 어려웠습니다.
  시각화 했을 때 상위 레이어로 갈 수록 이미지가 사람이 알아보기 어려운 형태로 변경되었는데,
  모델은 어떻게 더 많은 특징을 파악할 수 있었던 것인지 이해하는 데 좀 어려웠던 것 같습니다.

## 💬 질문 또는 토론 유도
- 하위 레이어에서는 Edge를 검출하는 필터, 상위 레이어에서는 Shape를 검출하는 필터를 사용한다고 배웠는데,
  하위 레이어와 상위 레이어에서 학습되는 필터는 실제로 어떤 방식이나 형태로 구성되며, 그 차이가 어떻게 나타나는지 궁금합니다.

**Comment by LeeHyunJong** [comment]
CNN의 구성요소에 대해서 잘 정리해주셨습니다. CNN의 연산과정에 대해서도 정리해두면 좋을 것 같습니다.

> 상위 레이어로 갈수록 Receptive Field를 넓게 보고, 더 많은 특징들을 파악할 수 있다는 부분이 어려웠습니다. 시각화 했을 때 상위 레이어로 갈 수록 이미지가 사람이 알아보기 어려운 형태로 변경되었는데, 모델은 어떻게 더 많은 특징을 파악할 수 있었던 것인지 이해하는 데 좀 어려웠던 것 같습니다. 하위 레이어에서는 Edge를 검출하는 필터, 상위 레이어에서는 Shape를 검출하는 필터를 사용한다고 배웠는데, 하위 레이어와 상위 레이어에서 학습되는 필터는 실제로 어떤 방식이나 형태로 구성되며, 그 차이가 어떻게 나타나는지 궁금합니다.

예를 들어, 3x3 필터를 사용하는 2층 CNN을 생각해보겠습니다. 첫 번째 레이어의 각 셀은 원본 이미지의 3x3 영역을 보고 특징을 추출합니다. 두 번째 레이어의 각 셀은 첫 번째 레이어의 3x3 영역을 보는데, 첫 번째 레이어 자체가 원본 이미지의 3x3을 반영하므로 결과적으로는 원본 이미지의 5x5 영역을 보게 됩니다. 이런 식으로 레이어가 깊어질수록 receptive field가 커져서, 모델은 더 넓은 맥락을 고려할 수 있습니다. 아래 이미지를 참조 바랍니다.

<img width="695" height="545" alt="Image" src="https://github.com/user-attachments/assets/bd9d0352-967c-4c9b-92a5-958d2b188c39" />

하위 레이어는 receptive field가 작기 때문에 주로 *엣지, 선, 색상 변화* 같은 단순한 패턴을 파악합니다. 그래서 시각화했을 때 이미지 윤곽이 도드라져 사람이 알아볼 수 있는 형태로 느껴집니다. 반대로, 상위 레이어로 갈수록 receptive field가 커지며 *이전 레이어의 단순한 패턴을 조합하여 더 큰 구조 또는 복잡한 패턴*(예: 여러 엣지 조합 → 특정 모양)을 파악합니다. 이때는 사람이 보기엔 잡음처럼 보여도, 모델 입장에서는 특정 클래스에만 잘 나타나는 유용한 신호가 됩니다.

CNN 필터는 사람이 인식 가능한 시각적 의미를 목표로 학습되는 것이 아니라, 분류·인식에 도움이 되는 (통계적) 패턴을 자동으로 학습하기 때문에 사람이 보기엔 점점 이해하기 어려운 모양이 되는 것입니다.

CNN 필터의 학습 과정을 직접 체험해보고 싶으시다면 [CNN Explainer](https://poloclub.github.io/cnn-explainer/) 같은 시각화 도구를 참고해 보시면 좋을 것 같습니다.

**Comment by jphan32** [comment]
## 보완 사항
- **Leaky ReLU**의 음수 구간 기울기를 0.1로 고정할 필요는 없습니다. 구현체마다 기본값이 0.01이거나 학습 가능한 파라미터(`PReLU`)로 설정되니, “작은 양수(예: 0.01)”처럼 범위를 열어 두면 좋겠습니다.
- **Batch Normalization**은 항상 Conv 다음, Activation 이전에만 오는 것이 아닙니다. ResNet 계열처럼 `Conv → BN → ReLU` 순서를 쓰는 경우도 있고, 일부 구현은 ReLU 이후에 배치하기도 하니 레이어 순서를 명시해 주세요.

## 보충 설명
- 상위 레이어로 갈수록 receptive field가 넓어진다는 말은 그 레이어의 한 픽셀이 원본 이미지에서 더 큰 영역을 바라본다는 뜻입니다. 초반 레이어는 에지나 색 대비 같은 작은 패턴에 반응하고, 깊은 레이어는 여러 작은 패턴을 조합해 윤곽을 잡습니다.

## 질문에 대한 답변
- 학습된 필터를 시각화해 보면 하위 레이어는 방향성 에지, 색 대비처럼 단순한 패턴이고, 상위 레이어는 여러 채널이 합쳐져 복잡한 질감이나 형태를 나타냅니다. `torchvision.utils.make_grid`나 TensorBoard filter 시각화를 사용해 레이어별로 직접 확인해 보세요.

## 학습 팁
- CIFAR 같은 작은 데이터셋으로 단일 Conv 레이어를 학습시키고 필터를 바로 시각화해 보면 이론이 빨리 체화됩니다.
- BatchNorm과 Dropout을 켰다 껐다 하면서 학습 곡선을 비교해 보세요. 안정화 효과가 눈에 띄게 바뀝니다.
- Pooling 대신 stride convolution을 적용해 보고 feature map 크기와 정확도를 같이 기록해 두면 구조 선택이 쉬워집니다.
